### ISSUE TITLE
[Enhancement]: Block-based checkout - update the Place order button

### ISSUE THREAD

**Issue by ralucaStan** [woothemes/woocommerce#46010]
This is a subticket of https://github.com/woocommerce/woocommerce/issues/44835, which has more context on the changes

- [ ]  Align with WooPay by adding a price to the purchase button. This is helpful especially when the sticky summary is not yet implemented

![CleanShot 2024-03-04 at 15 19 18@2x](https://github.com/woocommerce/woocommerce/assets/99638528/2ead8edf-98de-4d29-bdeb-c2c6f7bf7f9c)

> [!IMPORTANT]
> We already have the `placeOrderButtonLabel` filter in place, which enables the modification of the "Place order" button label. We should update this filter to also allow merchants to optionally remove the total price from the button. Additionally, it is essential to test this implementation against the use of existing filters to ensure backwards compatibility.

**Comment by nielslange** [comment]
@ralucaStan What should happen in the case of a free order? Shall we leave the price out, or shall we show £0.00? These are examples of https://woo.com/checkout/:

<table>
<tr>
<td valign="top">Free order:
<br><br>
<img width="340" alt="Screenshot 2024-03-28 at 15 04 23" src="https://github.com/woocommerce/woocommerce/assets/3323310/9795c720-6cf8-4ffa-84c3-5b27f539faf6">
</td>
<td valign="top">Paid order:
<br><br>
<img width="345" alt="Screenshot 2024-03-28 at 15 03 06" src="https://github.com/woocommerce/woocommerce/assets/3323310/8641698e-a44c-4b78-922b-9d2ef57c369f">
</td>
</tr>
</table>

**Comment by Barnabas2** [comment]
Better set it to "Free" instead of £0.00

**Comment by dominique-vias** [comment]
@nielslange Just make it say `Place order`

**Comment by nielslange** [comment]
> @nielslange Just make it say `Place order`

Thanks, @nikkivias. This makes the implementation pretty straightforward.

**Comment by nielslange** [comment]
@pmcpinto, @ralucaStan, and @nikkivias, during the implementation of this feature, I've run into a challenge related to the formatting of the place order button label, specifically:

```
{ LABEL } { DIVIDER } { PRICE }
```

As per the design, `·` is used as the divider. However, as highlighted by @nikkivias in https://github.com/woocommerce/woocommerce/issues/46010#issuecomment-2034562337, the divider and price should not be displayed when the total price is $0.00. This raises two important issues:

1. **Dynamic divider display**: How can we dynamically detect and hide the divider when the price is zero? This is crucial to ensure that we do not display unnecessary or misleading information.
2. **Merchant flexibility**: Should we allow merchants to customise the divider? If so, we need to determine the most suitable method to implement this flexibility.

Given these challenges, I have outlined several potential solutions for managing the divider.

<details>
<summary>Option 1: Fixed divider</summary>
<br>

This approach uses a hardcoded divider. The default label is `Place order · <price/>`. If the total price is $0.00, we revert to `Place order`.

**Benefits:**

- Simple and easy to implement.

**Drawbacks:**

- No flexibility for merchants to customise the divider.
<hr>
</details>

<details>
<summary>Option 2: CSS class for divider</summary>
<br>

We can wrap the divider in a CSS class (e.g., `.wc-block-components-checkout-place-order-button-price-divider`), allowing merchants to customise it via CSS:

```css
.wc-block-components-checkout-place-order-button-price-divider {
  content: " - ";
}
```

If the total price is $0.00, we're not rendering the corresponding HTML element.

**Benefits:**

- Easy to implement and allows for styling customization.

**Drawbacks:**

- Adds minimal HTML/CSS overhead.
- Customization requires CSS knowledge, which may not be intuitive for all merchants.
<hr>
</details>

<details>
<summary>Option 3: Editable divider setting</summary>
<br>

Introduce a setting in the place order component, allowing merchants to choose their preferred divider from a dropdown menu. If the total price is $0.00, we're not rendering the selected divider.

**Benefits:**

- Directly configurable within the WooCommerce settings, enhancing usability for merchants.

**Drawbacks:**

- More complex to implement compared to other solutions.
- Merchants can only select from available dividers, but cannot define custom dividers.
- This solution does not work when using the `placeOrderButtonLabel` filter.
<hr>
</details>

<details>
<summary>Option 4: Whitelist common dividers</summary>
<br>

Merchants can select from a list of whitelisted dividers by updating the button label in the editor or applying the `placeOrderButtonLabel` filter. If the total price is $0.00, we're removing all whitelisted characters. 

**Benefits:**

- Offers flexibility within a controlled range of options.
- Simple to implement.

**Drawbacks:**

- Requires clear documentation to inform merchants about the available options.

<hr>
</details>

<details>
<summary>Option 5: Allow all characters and use regex</summary>
<br>

Merchants can use all characters as dividers. If the total price is $0.00, we're removing everything that is not a character `[a-zA-Z]` or a space `[\s]` using regex. The implementation for this scenario could look like this:

```js
if ( label.includes( '<price/>' ) ) {
	if ( cartTotals.total_price === '0' ) {
		label = label.replace( '<price/>', '' );
		label = label.replace( /[^a-zA-Z\s]/g, '' );
	} else {
		label = label.replace( '<price/>', totalPrice );
	}
}
```

**Benefits:**

- Offers the highest possible flexibility, besides the CSS class approach.
- Simple to implement.

**Drawbacks:**

- Merchants might use characters that could lead to unwanted side effects.

<hr>
</details>

I'm looking forward to hearing your thoughts on the potential solutions. Personally, I'm leaning towards option 5, as it offers the highest flexibility while not polluting the HTML with additional elements.

**Comment by pmcpinto** [comment]
@nielslange thanks for showcasing different options to solve this problem. How would the merchant customize the divider with option 5? Is it just through a filter?

**Comment by nielslange** [comment]
> @nielslange thanks for showcasing different options to solve this problem. How would the merchant customize the divider with option 5? Is it just through a filter?

Thanks for asking, @pmcpinto. Option 5 works both for editing the place order button label in the editor and via the `placeOrderButtonLabel` filter. The screenshots below show how the merchant could customise the divider. 

> [!IMPORTANT]
> The steps and screenshots below show my WIP branch and cannot be reproduced on `trunk`. Both the steps and the screenshots are for demo purposes only.

**Editing the place order button label in the editor**

1. Create a test page and add the Checkout block to it.
2. Scroll down to the `Place Order` button.
3. Click into the button.
4. Edit the text, e.g. `Pay now · <price/>`.

<table>
<tr>
<td valign="top">With non-zero price:
<br><br>
<img width="2560" alt="Screenshot 2024-04-30 at 19 49 21" src="https://github.com/woocommerce/woocommerce/assets/3323310/0f2c953a-5721-4d6a-bccb-d3e65a295128">
</td>
<td valign="top">With zero price:
<br><br>
<img width="2560" alt="Screenshot 2024-04-30 at 19 51 09" src="https://github.com/woocommerce/woocommerce/assets/3323310/6d217839-10e7-4a00-ba52-bdb3aefbde04">
</td>
</tr>
</table>

**Editing the place order button label via the `placeOrderButtonLabel` filter**

1. Create a test page and add the Checkout block to it.
2. Install and activate the [Simple Custom CSS and JS](https://wordpress.org/plugins/custom-css-js/) plugin.
3. Go to `/wp-admin/post-new.php?post_type=custom-css-js&language=js`.
4. Add the following code snippet:

```js
const { registerCheckoutFilters } = window.wc.blocksCheckout;

const modifyPlaceOrderButtonLabel = ( defaultValue, extensions ) => {
	return 'Purchase now: <price/>';
};

registerCheckoutFilters( 'example-extension', {
	placeOrderButtonLabel: modifyPlaceOrderButtonLabel,
} );
```

<table>
<tr>
<td valign="top">With non-zero price:
<br><br>
<img width="2560" alt="Screenshot 2024-04-30 at 19 46 42" src="https://github.com/woocommerce/woocommerce/assets/3323310/64dc366c-a7df-4e2b-850c-cb1cae163b7b">
</td>
<td valign="top">With zero price:
<br><br>
<img width="2560" alt="Screenshot 2024-04-30 at 19 44 14" src="https://github.com/woocommerce/woocommerce/assets/3323310/5cb33ff2-b1c8-46b1-a44f-29392f0f77b6">
</td>
</tr>
</table>

**Comment by pmcpinto** [comment]
@nielslange thanks for the additional context.

Adding the customization via the `placeOrderButtonLabel` filter makes total sense, as it's possible to update the button label and we also plan to allow users to remove the price. 

In the editor, I think we should only enable the divider customization if we also do the same for the label and pricing display. Do you agree? Are there any risks we should consider? cc @nikkivias @elizaan36 for feedback.

**Comment by nielslange** [comment]
Thanks for your reply, @pmcpinto. 

> In the editor, I think we should only enable the divider customization if we also do the same for the label and pricing display.

Can you elaborate on that? I'm afraid, I don't understand what you mean by that. 🙈

**Comment by ralucaStan** [comment]
The `Place order · <price/>` comes as a default on the button; merchants can modify that label in the Editor, by interacting with the text and developers can use the filter to replace that text all together. 

I appreciate the thought that went into this but having a specific setting for the divider seems like an overkill to me; I would rephrase the questions to: 

- What should the filter replace?
- What should the merchant pe able to edit in the editor?

Do we want to display the price also for buttons that are editable by the filter? Or do we allow 3PDs to display whatever they want? I imagine one can say `Pay with X` case in which the price does not make sense. 

Do we want the merchat to be able to edit the default we offer? Can they remove the price? Can they only edit the text around the price?

Determining the amount of flexibility we want to allow will help up with picking a technical solution.

**Comment by nielslange** [comment]
> The `Place order · <price/>` comes as a default on the button; merchants can modify that label in the Editor, by interacting with the text and developers can use the filter to replace that text all together.

That's how I've currently implemented it.

> I appreciate the thought that went into this but having a specific setting for the divider seems like an overkill to me; 

That's what I'm thinking as well.

> I would rephrase the questions to:
> 
> * What should the filter replace?
> * What should the merchant pe able to edit in the editor?

Would rephrasing the questions change the expected outcome? Allow me to elaborate by answering your questions:

- What should the filter replace? → The `placeOrderButtonLabel` filter allows change the label of the "Place order" button. Merchants could use the placeholder `<price>`, e.g. `Pay now · <price/>`. However, the `placeOrderButtonLabel` filter itself does not replace anything. It would simply display what the merchant defined. The question is what if the merchant used the `<price/>` placeholder and the order has a zero price. As we don't want to display the zero price, we should not display the divider either.
- What should the merchant pe able to edit in the editor? → Similar to my previous answer. By default, the place order button label is `Place order · <price/>`. A merchant could change it to `Pay now · <price/>` or to `Buy now`. In the first case, we'd need to strip out the price and the divider, while in the second case, we could show what the merchant defined, given that no `<price/>` placeholder has been provided.

> Do we want to display the price also for buttons that are editable by the filter? Or do we allow 3PDs to display whatever they want? I imagine one can say `Pay with X` case in which the price does not make sense.

The `placeOrderButtonLabel` filter only controls the display logic of the place order button label. This PR does not alter the display logic for 3PDs. Whatever they define, will get displayed. No `<price/>` placeholder gets injected in their call.

> Do we want the merchat to be able to edit the default we offer? Can they remove the price? Can they only edit the text around the price?

By default (for new installations), the place order button label would be `Place order · <price/>`. If merchants would rather not display the price, they can alter the place order button label to `Place order`, bot in the editor or via the `placeOrderButtonLabel` filter. That way, merchants have full control over whether the price should be displayed on the button or now.

**Comment by pmcpinto** [comment]
> Can you elaborate on that? I'm afraid, I don't understand what you mean by that. 🙈

Sure. Right now, we don't offer a way to edit the button label. It only makes sense to allow them to change the divider, if we also allow changing the label and removing the price, which is what I think you're suggesting.

**Comment by pmcpinto** [comment]
Niels, my bad. I forgot that it was already possible to change the label. Thanks for the clarification. In that case, we should keep it consistent and allow users also to change the divider and remove the price. cc @nikkivias

**Comment by nielslange** [comment]
Thanks for your reply, @pmcpinto. Shall we proceed with option 5 as mentioned in https://github.com/woocommerce/woocommerce/issues/46010#issuecomment-2084740443? It's easy to implement and provides the highest flexibility.

**Comment by dominique-vias** [comment]
@nielslange option 5 looks good to me, thank you for showing all possibilities.

I think a general guideline we can use with these new blocks changes is that we are _only prescribing smart defaults_, but we must be able to provide flexibility as much as is possible, and as long as it doesn't prove harmful.

**Comment by pmcpinto** [comment]
> @pmcpinto. Shall we proceed with option 5 as mentioned in https://github.com/woocommerce/woocommerce/issues/46010#issuecomment-2084740443? It's easy to implement and provides the highest flexibility.

Yes. Unless @ralucaStan still has some concerns about this approach.

**Comment by ralucaStan** [comment]
We've reverted the initial approach for fixing this issue. Instead of reopening this I've created a new issue https://github.com/woocommerce/woocommerce/issues/48433 that is currently blocked as it needs design and a discussion about the next steps.

@elizaan36 for awarness
